Adding A Modal Component

aryanpingle opened the pull request:
This PR adds a custom Modal component so that Squoosh can communicate better with users (_snackbars are SO last year_).

While the snackbar popup will convey small and concise messages to users, the Modal can be used for displaying verbose  information like:

1. **INFO** mode - Explaining codecs and their parameters
2. **UPDATE** mode - Informing users about new features
3. **ERROR** mode - Expanding error messages on mobile, where users can't view the console error messages

It's keyboard accessible, has pre-formatted styles for different HTML elements, is responsive across the board and can be easily accessed by other components/elements.

I've tested the feature on the following devices:

* Safari - iOS / Mac
* Chrome - iOS / Android / Windows
* Firefox - Android / Windows

---

### DEMO

In the right-side options under the quality slider, click on any of the modal checkboxes. If you're using a keyboard, focus will be restored to that checkbox once the modal is closed.

---

**Unnecessary PR Lore**: This is complementary to PR #1366 where Surma rightly pointed out that the HTML hover _title_ is inaccessible to mobile users. I initially planned to put small one-line descriptions for each codec param in the _title_ attribute, then later add modal support with more detailed descriptions. But I figure if those one-liners will be overwritten anyway, might as well do them soon.

aryanpingle committed:
Add modal skeleton

aryanpingle committed:
Improve the modal component

* Improvements to Modal UI
* Modal is now keyboard accessible
* Added new/better info, error and update icons

aryanpingle committed:
Change Modal directory + color scheme

aryanpingle commented:
@argyleink @surma could y'all take a look at this whenever you're free?

surma commented on src/client/lazy-app/Compress/Output/index.tsx:
```diff
@@ -331,6 +331,18 @@ export default class Output extends Component<Props, State> {
+            {Math.round(scale * originalImage.width)} ×{' '}
+            {Math.round(scale * originalImage.height)}
+          </div>
+        )}
```
Is this supposed to be part of the PR? I think this can be really useful but pretty independent of the feature :D

surma replied:
<img width="769" alt="Screenshot 2023-06-20 at 10 52 17" src="https://github.com/GoogleChromeLabs/squoosh/assets/234957/12eced32-c18b-40ba-b567-2d84b22d7c2d">
The text can also get quite easily lost due to contrast.

surma commented on src/client/lazy-app/Compress/Modal/index.tsx:
```diff
@@ -0,0 +1,148 @@
+      shown: true,
+    });
+    // Wait for the 'display' reset to take place, then focus
+    setTimeout(() => {
```
```suggestion
    requestAnimationFrame(() => {
```

surma replied:
But on second thought, it might actually be more idiomatic to use the `ref` prop to focus the button once a DOM element been created for it. Something like

```js
  <button ref={(el) => el?.focus()} /*...*/>
```

aryanpingle replied:
I'm confused, will that work in this case? I'm not adding/removing the Modal component to the DOM tree on show/hide, just changing some CSS properties to "hide" it from the user.

jakearchibald replied:
I think the correct Preact way to do this would be in `componentDidUpdate`. But, you get all this stuff for free with `<dialog>`, so we should just use that.

surma commented on src/client/lazy-app/Compress/Modal/index.tsx:
```diff
@@ -0,0 +1,148 @@
+  }
+
+  private _getCloseButton() {
+    return this.modal!.querySelector('button')!;
```
We should have a dedicated class or id for the close button. Feels brittle otherwise with user-provided content for the modal.

aryanpingle replied:
That was just lazy coding on my part LOL, I'll make a dedicated id for it

jakearchibald replied:
Also seems odd to have a method like this, but then other code above getting the button manually. But I don't think we'll need this at all with `<dialog>`.

surma commented on src/client/lazy-app/Compress/Modal/index.tsx:
```diff
@@ -0,0 +1,148 @@
+    return focusables[focusables.length - 1] as HTMLElement;
+  }
+
+  private _onKeyDown(e: KeyboardEvent) {
+    // If Escape, hide modal
+    if (e.key === 'Escape' || e.keyCode == 27) {
+      this.hideModal();
+      e.preventDefault();
+      e.stopImmediatePropagation();
+      return;
+    }
+
+    let isTabPressed = e.key === 'Tab' || e.keyCode === 9;
+
+    if (!isTabPressed) return;
+
+    if (e.shiftKey) {
+      // If SHIFT + TAB was pressed on the first focusable element
+      // Move focus to the last focusable element
+      if (document.activeElement === this._getCloseButton()) {
+        this._getLastFocusable().focus();
+        e.preventDefault();
+        e.stopImmediatePropagation();
+      }
+    } else {
+      // If TAB was pressed on the last focusable element
+      // Move focus to the first focusable element
+      if (document.activeElement === this._getLastFocusable()) {
+        this._getCloseButton().focus();
+        e.preventDefault();
+        e.stopImmediatePropagation();
+      }
+    }
+  }
```
On the one hand, I am impressed that you managed to successfully trap focus manually :D

However, I think there are some platform features that we can use to achieve the same thing with less code. The first options is the `inert` attribute, that can make all other HTML in the tree unfocusable. Read [more here](https://developer.chrome.com/articles/inert/). The other option, and the even more natural fit I think, is using the `<dialog>` element, which is now [well supported](https://caniuse.com/dialog).

aryanpingle replied:
I **did** want to use the `<dialog>` element, but isn't it fairly new on iOS? I personally know lots of people using iOS 14 and who stubbornly refuse to update. But yeah, most user devices should be able to support it, and I'm all for using a cool new web feature. Lemme know what you think!

aryanpingle replied:
I was iffy about using `inert` because I'd have to apply it to all element siblings of the Modal components. Wouldn't that initiate a huge reflow every time the Modal is shown/hidden?

surma commented on src/client/lazy-app/Compress/Modal/style.css:
```diff
@@ -0,0 +1,174 @@
+  gap: 0.5em;
+  border-bottom: 1px solid rgba(0, 0, 0, 8%);
+
+  & > * {
+    height: 100%;
+    font-size: 1.5em;
+  }
```
This seems a bit dangerous. I think instead of setting a height on each child, setting `align-content: stretch;` on the header should do the trick. Same with the font size, which is automatically inherited.

surma commented on src/features/encoders/mozJPEG/client/index.tsx:
```diff
@@ -32,6 +36,46 @@ interface State {
+        </figcaption>
+      </figure>
+      <p>
+        That's a link.{' '}
```
lol

surma commented on src/client/lazy-app/Compress/Modal/modal-context.ts:
Considering that there can only ever be one modal shown at time (it’s a singleton), I wonder if we need to do the Context/Provider dance. We could just have a ES module that exports a `showModal()` function and forwards that to a `ref` of a single, global `<Modal>` instance. Not sure what others think of that, tho

jakearchibald replied:
I think this is another problem solved by `<dialog>`

surma commented on src/client/lazy-app/Compress/Modal/index.tsx:
```diff
@@ -0,0 +1,148 @@
+export interface ModalMessage {
+  type: 'info' | 'error' | 'update';
+  title: string;
+  content: string;
```
I don’t think `content` should be a string, but rather a `VNode`. With string we lose the ability to have more complex interactivity in the modal (event listeners would get lost etc). Even `title` should probably be a `VNode`.

The next logical step is to make the Modal itself actually unaware of whether it’s an error or info or update. It’s just a component for modal dialogs, and since we can pass arbitrary VNodes for title and content, the user can style it depending on what the use-case is.

aryanpingle replied:
Ahh you're right, `VNode` would be better suited. Will do!

jakearchibald replied:
Wouldn't it be more natural for the content to be the node's children? (like how `<dialog>` works)

surma reviewed (changes requested):
Great work! This is gonna be a super useful primitive to have in Squoosh going forward.

Have a couple of remarks, and a bigger rework request/idea around the `<dialog>` element. LMK what you think!

jakearchibald commented on src/client/lazy-app/Compress/Modal/index.tsx:
```diff
@@ -0,0 +1,148 @@
+import { h, Component } from 'preact';
```
This file is in `/Compress/` - is its usage intended to be scoped to this part of the app, or is it for more general use?

aryanpingle replied:
General use. Maybe I could put it in the shared directory like the snack bar element?

jakearchibald replied:
`lazy-app` is a good place for it if it isn't needed for the first interaction.

jakearchibald commented on src/client/lazy-app/Compress/Modal/index.tsx:
```diff
@@ -0,0 +1,148 @@
+interface Props {}
+
+export interface ModalMessage {
+  type: 'info' | 'error' | 'update';
```
Is the intent to use the 'type' for other things? Right now it's just impacting the icon. So, maybe accept a vnode for the icon?

aryanpingle replied:
I can't really think of reasons other than these 3 for displaying a modal - so I thought it might be better to abstract away the Icon behind the `type` parameter

jakearchibald replied:
The benefit of providing the icon is unused icons won't be bundled

jakearchibald commented on src/client/lazy-app/Compress/Modal/index.tsx:
```diff
@@ -0,0 +1,148 @@
+    shown: false,
+  };
+
+  private modal?: HTMLElement;
+  private returnFocusElement?: HTMLElement | null;
```
Not required, but we can start using 'real' private properties. If you want to do that, you might need to adjust the config.

The benefit in real private vars is their names compress.

aryanpingle replied:
Ahh got it, will do

jakearchibald reviewed (changes requested):
I didn't do a full review, as I think this will change pretty significantly when it moves to use `<dialog>`, which will hopefully reduce the complexity.

argyleink commented on src/client/lazy-app/Compress/Modal/style.css:
```diff
@@ -0,0 +1,174 @@
+  pre {
+    overflow-x: auto;
+  }
+}
```
```suggestion
  hr {
    border: none;
    border-bottom: 1px solid var(--dim-text);
  }
}
```

argyleink commented on src/client/lazy-app/Compress/Modal/style.css:
```diff
@@ -0,0 +1,174 @@
+
+  pre,
+  code {
+    background-color: rgba(0, 0, 0, 10%);
```
```suggestion
    background-color: var(--black);
```

argyleink commented on src/client/lazy-app/Compress/Modal/style.css:
```diff
@@ -0,0 +1,174 @@
+  }
+
+  pre {
+    overflow-x: auto;
```
```suggestion
    overflow-x: auto;
    padding: 1rem 2rem;
```

argyleink commented on src/client/lazy-app/Compress/Modal/style.css:
```diff
@@ -0,0 +1,174 @@
+  line-height: 1.5;
+
+  padding: 0 1em;
+
```
```suggestion
  h1 {
    color: var(--white);
  }
```

argyleink commented on src/client/lazy-app/Compress/Modal/style.css:
```diff
@@ -0,0 +1,174 @@
+  max-height: 100%;
+  padding: 1em;
+
+  background-color: rgba(0, 0, 0, 6%);
```
```suggestion
  background-color: var(--dark-gray);
```

argyleink commented on src/client/lazy-app/Compress/Modal/style.css:
```diff
@@ -0,0 +1,174 @@
+  max-width: 60ch;
+  max-height: 70vh;
+
+  border: 4px solid black;
```
```suggestion
```

argyleink commented on src/client/lazy-app/Compress/Modal/style.css:
```diff
@@ -0,0 +1,174 @@
+  border-radius: 10px;
+  overflow: hidden;
+
+  background-color: hsl(250deg, 100%, 90%);
```
```suggestion
  background-color: var(--off-black);
  color: var(--light-gray);
```

argyleink commented on src/client/lazy-app/Compress/Modal/style.css:
```diff
@@ -0,0 +1,174 @@
+
+  a {
+    &:link {
+      color: hsl(220, 100%, 50%);
```
```suggestion
      color: var(--blue);
```

argyleink commented on src/client/lazy-app/Compress/Modal/style.css:
```diff
@@ -0,0 +1,174 @@
+      color: hsl(220, 100%, 50%);
+    }
+    &:visited {
+      color: rgb(255, 0, 103);
```
```suggestion
      color: var(--pink);
```

argyleink reviewed (changes requested):
Overall pretty good! Spacing works, sizes work, but the colors were a bit off the rails, so I've made a few suggestions to bring it back into the theme.

<img width="745" alt="Screenshot 2023-06-26 at 1 46 57 PM" src="https://github.com/GoogleChromeLabs/squoosh/assets/1134620/2e520a91-18bc-4cdf-9172-908739d13e4f">

A couple other notes!
- I agree with others that dialog is a good choice here, it'll save a bunch of maintenance burden (like trying to trap focus, etc) and I dont think a majority of the users are on iOS making these transformations
- The labels should probably also be keyboard focusable. It's a bit strange that focus is "restored" but not really to where I came from, rather the nearest checkbox to where I came from. Also not sure how to invoke this from keyboard, tho it's clear how to dismiss it.

Cool idea here. Hope this review is helpful!

aryanpingle commented:
@argyleink That's a wayyy better design, thank you for the review!

And yep, my plan is to give an optional prop to the `Range` / `Checkbox` / etc components that turns their labels into modal-activating keyboard-navigable buttons. So far this was just a test to see if focus can be returned to the calling checkbox, which it can.

Working on a refactor using the `<dialog>` element as we speak 👍

aryanpingle committed:
Bump TS version for `<dialog>` support

TS 4.8.3 onwards supports `showModal()`. Also bumped @types/node.

aryanpingle committed:
Refactor modal to use `<dialog>`

aryanpingle committed:
Modal - use VNode icon and content

aryanpingle commented:
- [x] Placed the Modal component directory in `lazy-app` instead of in `Compress`
- [x] Refactored using the `<dialog>` element
- [x] Used the improved design by @argyleink 
- [x] Switched to using `VNode`s for modal content and icon
- [x] Switched from using Context/Provider to shared functions 

Caveats:

* ~~Still uses Context/Provider to share `showModal()`. Can't use props like with `showSnack` because deeply nested children should be able to access it. Gonna try @surma 's suggestion of a single global `showModal()` function that can use the modal as soon as it's mounted.~~
* The CSS for fading the modal in and out is a bit ugly, but I couldn't find an easier way considering the dialog element's nuances.

aryanpingle commented:
Switched from using Context/Provider to shared functions

surma commented on src/client/lazy-app/Modal/index.tsx:
```diff
@@ -0,0 +1,128 @@
+    // Make the modal fade out
+    this.dialogElement.classList.add(style.modalClosing);
+
+    this.setState({
+      message: { ...this.state.message },
+      shown: false,
+    });
```
```suggestion
  this.setState(cleanSet(this.state, `shown`, false));
```

aryanpingle replied:
TIL about `cleanSet`. On it!

surma commented on src/client/lazy-app/Modal/index.tsx:
```diff
@@ -0,0 +1,128 @@
+  private _onKeyDown(e: KeyboardEvent) {
+    // Default behaviour of <dialog> closes it instantly when you press Esc
+    // So we hijack it to smoothly hide the modal
+    if (e.key === 'Escape' || e.keyCode == 27) {
```
I [don’t think](https://developer.mozilla.org/en-US/docs/Web/API/KeyboardEvent#:~:text=Toggle%20history-,key,-51) we need both.

```suggestion
    if (e.key === 'Escape') {
```

surma commented on src/client/lazy-app/Modal/modal-context.ts:
I think this file is unused?

aryanpingle replied:
Nice catch, I'll remove it from the commits and do some cleanup

surma commented on src/client/lazy-app/Modal/index.tsx:
```diff
@@ -0,0 +1,128 @@
+  }
+
+  private _showModal(message: ModalMessage) {
+    if (!this.dialogElement) return;
```
It might be worth throwing here (something like “called `Modal.showModal()` without putting the `<Modal />` component being present”).

aryanpingle replied:
Sure thing!

surma reviewed (approved):
This looks really good!

Some minor nits, and of course you should remove the dummy checkboxes and content before merging, but otherwise LGTM! 🎉

aryanpingle committed:
Replace Context with shared functions

argyleink reviewed (approved):
approved but with 1 change request!
- [x] the mobile layout has the dialog at `translateY(0)`, doesnt look right: pop that to `20px` or so it looks more intentional 👍🏻

aryanpingle commented on src/client/lazy-app/Modal/style.css:
```diff
@@ -0,0 +1,200 @@
+
+dialog[open] {
+  opacity: 1;
+  transform: translateY(0px);
```
@argyleink I'm a bit confused, you suggested 20px to show intent - but wouldn't it be better to add a comment here instead of changing the translation amount? 20px isn't making much of a visual difference on mobile IMO

argyleink replied:
sorry for the confusion!

this
![Screenshot 2023-07-10 at 12 20 45 PM](https://github.com/GoogleChromeLabs/squoosh/assets/1134620/059873d0-2ea6-4739-a7e6-b8b987ce15fb)

is better than this
![Screenshot 2023-07-10 at 12 20 33 PM](https://github.com/GoogleChromeLabs/squoosh/assets/1134620/8b730d3d-9894-4078-bc27-805bab41a262)

is what I'm getting at. the intent is to have a modal, and it's weird if it's stuck to the top flush, better if it's got some breathing room from the top edge.

aryanpingle replied:
Huh, that's weird - it's supposed to be centered on all devices/sizes, not stuck to the top. Could you tell me which browser/os you're using? I've tested it on most Android browsers and remotely tested on a couple iOS versions too, seemed to work fine on those.

argyleink replied:
Chrome stable 114

argyleink replied:
confirm devtools issue

jakearchibald commented on src/client/lazy-app/Modal/index.tsx:
```diff
@@ -0,0 +1,127 @@
+      icon: <svg></svg>,
+      title: 'default',
+      content: <Fragment></Fragment>,
+    },
```
What's the benefit of having this as one object rather than individual properties.

jakearchibald replied:
Why are these state rather than props?

jakearchibald commented on src/client/lazy-app/Modal/index.tsx:
```diff
@@ -0,0 +1,127 @@
+  };
+
+  private dialogElement!: HTMLDialogElement;
+  static modalInstance?: Modal | undefined;
```
Does it need to be static? One of the benefits of `<dialog>` is that it can be deep in the DOM but still pop out to the top-level. I wonder if you're trying to recreate that feature here, but given it already exists in the platform, we can just use it.

jakearchibald commented on src/client/lazy-app/Modal/index.tsx:
```diff
@@ -0,0 +1,127 @@
+    this.dialogElement.addEventListener(
+      'transitionend',
+      this._closeOnTransitionEnd.bind(this),
+    );
```
I don't think this is the right way to do this. The `transitionend` even bubbles, so the wrong transition may end up closing the dialog. Also, if it somehow ends up trying to transition to the same value, this event won't trigger. No requirement to watch it, but there's a 203 that covers these gotchas https://www.youtube.com/watch?v=9-6CKCz58A8

If you want to do something after an animation, the safest way is to do the animation with the web animation API, and then use its `finished` promise.

jakearchibald commented on src/client/lazy-app/Modal/index.tsx:
```diff
@@ -0,0 +1,127 @@
+      'transitionend',
+      this._closeOnTransitionEnd.bind(this),
+    );
+    this.dialogElement.setAttribute('inert', 'enabled');
```
I think this stuff will be refactored anyway, but for future reference, you can just do:

```js
el.inert = true;
```

jakearchibald commented on src/client/lazy-app/Modal/index.tsx:
```diff
@@ -0,0 +1,127 @@
+    Modal.modalInstance = this;
+  }
+
+  private _closeOnTransitionEnd() {
```
fwiw, I'm happy for new code to use real JS privacy. As in `#closeOnTransitionEnd`

jakearchibald reviewed (changes requested):
This is heading in the right direction, but I think trying to make the modal a singleton is unnecessary. 

One of the benefits of `<dialog>` is that it doesn't need to be placed out-of-context in the top of the DOM, as it pops into the top-layer when it opens. It feels like the implementation will become simpler if we lean into that ("use the platform" etc etc).

So, the usage of this component would be:

```jsx
<Modal shown={bool} icon={iconJSX} title="Hello!">
  Modal content goes here.
</Modal>
```

We'd have many of those in our JSX, in context.

And when `shown` changes, the component handles the animated show and hide.

I guess at some point we'd also have a tooltip component that outputs an icon and a modal, and causes the modal to show when the icon is clicked.

jakearchibald commented:
Thinking about the helper component:

```jsx
<ModalHelpIcon title="Hello!">
  Modal content goes here.
</ModalHelpIcon>
```

So the above would output a ℹ️ icon (or whatever), and a `<Modal />`. When the icon is clicked, it shows the modal, by toggling the `shown` prop passed to `<Modal />`.

That means we can just add `<ModalHelpIcon />` where we want to details codecs etc, and that component takes care of the show/hide state.

aryanpingle committed:
Remove singleton pattern from modal

aryanpingle committed:
Add ModalHint

aryanpingle committed:
Basic `ModalHint` demo

aryanpingle commented:
Added a `ModalHint` component that contains a Modal within itself and shows it when activated. Went with a subtle green accent color so that hints like in the image below stand out. @argyleink, would love your thoughts on this!

![image](https://github.com/GoogleChromeLabs/squoosh/assets/49340972/c1d62610-284c-4fda-96d6-019d827ec9f2)

argyleink commented:
fun additions!

I've got a 2 overall comments 🙂 

1. replace the green with `var(--blue)` (maybe) or `var(--less-light-gray)` (prolly), like you've used in the dialog header here. info icons shouldn't try and steal attention, unless hovered, you can spice them up then.  
![Screenshot 2023-07-26 at 7 38 22 AM](https://github.com/GoogleChromeLabs/squoosh/assets/1134620/3f3ab8b5-5a27-49fd-9a68-82a73854d587)

2. there's alignment issues, here's some screenshots 
![Screenshot 2023-07-26 at 7 38 06 AM](https://github.com/GoogleChromeLabs/squoosh/assets/1134620/db83c588-2477-4106-9386-d8d19f939bab) ![Screenshot 2023-07-26 at 7 38 33 AM](https://github.com/GoogleChromeLabs/squoosh/assets/1134620/78c4deac-005a-4a33-8bf1-3f31918c47ea)

aryanpingle committed:
Add `animateTo` function

aryanpingle committed:
Switch to Web Animations API

aryanpingle committed:
Catch animation errors for ::backdrop

aryanpingle commented:
* Fixed `ModalHint` alignment issues
* Switched from transitionEnd event handler to the Web Animations API

jakearchibald commented on src/features/encoders/mozJPEG/client/index.tsx:
```diff
@@ -114,7 +163,9 @@ export class Options extends Component<Props, State> {
             onInput={this.onChange}
           >
-            Quality:
+            <ModalHint modalTitle="mozJPEG - quality" content={sampleContent()}>
+              Quality:
+            </ModalHint>
```
I think it makes more sense for the content of the modal to be the child of `<ModalHint>`, and the 'label' to be a prop. The label could probably just be a string too.

aryanpingle replied:
Hmm I suppose that makes more sense. My thought process was we can just wrap existing text nodes (like for the custom input elements) in the `ModalHint` and get the modal functionality without much refactoring - but in hindsight that's a cop-out.

jakearchibald replied:
It works both ways, and in this case it isn't clear-cut which thing is naturally the "children". But "content" sounds like just another name for "children", whereas the actual child is the label. That, and I usually lean towards having the larger, more tree-like content as the child.

But yeah, it isn't clear-cut.

jakearchibald commented on src/client/lazy-app/icons/index.tsx:
```diff
@@ -11,6 +11,36 @@ const Icon = (props: preact.JSX.HTMLAttributes) => (
+      clip-rule="evenodd"
+      d="M64 0C28.6538 0 0 28.6538 0 64V448C0 483.346 28.6538 512 64 512H448C483.346 512 512 483.346 512 448V64C512 28.6538 483.346 0 448 0H64ZM297.543 377.307C298.628 376.1 299.756 374.845 300.942 373.528C305.12 368.875 306.716 364.537 304.731 357.714C302.602 350.359 293.989 346.698 286.361 349.472C276.039 353.254 258.652 363.899 258.652 363.899C258.652 363.899 271.846 302.925 272.394 300.272C27 ...
+    />
+  </svg>
```
Passed it through SVGOMG.

```suggestion
  <svg viewBox="0 0 512 512" fill="currentColor">
    <path fill-rule="evenodd" d="M64 0C29 0 0 29 0 64v384c0 35 29 64 64 64h384c35 0 64-29 64-64V64c0-35-29-64-64-64H64Zm234 377 3-3c4-5 6-9 4-16-2-8-11-11-19-9-10 4-27 15-27 15l13-64 4-15c6-25 14-53-4-72-15-16-34-12-51 4l-24 25-3 4c-4 5-6 9-4 16s11 11 18 8c11-3 28-14 28-14l-12 56-1 1c-3 12-20 73-1 94 15 16 35 12 51-4l25-26Zm-72-240a43 43 0 1 0 87-8 43 43 0 0 0-87 8Z"/>
  </svg>
```

jakearchibald commented on src/client/lazy-app/icons/index.tsx:
```diff
@@ -11,6 +11,36 @@ const Icon = (props: preact.JSX.HTMLAttributes) => (
+      clip-rule="evenodd"
+      d="M12 22c5.523 0 10-4.477 10-10S17.523 2 12 2 2 6.477 2 12s4.477 10 10 10zm-1.5-5.009c0-.867.659-1.491 1.491-1.491.85 0 1.509.624 1.509 1.491 0 .867-.659 1.509-1.509 1.509-.832 0-1.491-.642-1.491-1.509zM11.172 6a.5.5 0 0 0-.499.522l.306 7a.5.5 0 0 0 .5.478h1.043a.5.5 0 0 0 .5-.478l.305-7a.5.5 0 0 0-.5-.522h-1.655z"
+    />
+  </svg>
```
```suggestion
  <svg viewBox="0 0 24 24" fill="currentColor">
    <path
      fill-rule="evenodd"
      d="M12 22c5.523 0 10-4.477 10-10S17.523 2 12 2 2 6.477 2 12s4.477 10 10 10zm-1.5-5.009c0-.867.659-1.491 1.491-1.491.85 0 1.509.624 1.509 1.491 0 .867-.659 1.509-1.509 1.509-.832 0-1.491-.642-1.491-1.509zM11.172 6a.5.5 0 0 0-.499.522l.306 7a.5.5 0 0 0 .5.478h1.043a.5.5 0 0 0 .5-.478l.305-7a.5.5 0 0 0-.5-.522h-1.655z"
    />
  </svg>
```

jakearchibald commented on src/client/lazy-app/icons/index.tsx:
```diff
@@ -11,6 +11,36 @@ const Icon = (props: preact.JSX.HTMLAttributes) => (
+    xmlns="http://www.w3.org/2000/svg"
+  >
+    <path d="M240 480C240 360 120 240 0 240C120 240 240 120 240 0C240 120 360 240 480 240C360 240 240 360 240 480Z" />
+  </svg>
```
```suggestion
  <svg viewBox="0 0 48 48" fill="currentColor">
    <path d="M24 48c0-12-12-24-24-24 12 0 24-12 24-24 0 12 12 24 24 24a25.8 25.8 0 0 0-24 24Z"/>
  </svg>
```

jakearchibald commented on src/client/lazy-app/Modal/style.css:
```diff
@@ -0,0 +1,171 @@
+dialog {
```
Instead of styling all dialogs, use a class name. That avoids clashes with other dialogs we may have in future.

jakearchibald commented on src/client/lazy-app/Modal/style.css:
```diff
@@ -0,0 +1,171 @@
+  display: grid;
+  outline-color: var(
+    --medium-light-gray
+  ) !important; /* Overwrite default black outline */
```
Where's this default coming from? As in, why does it need `!important`?

aryanpingle replied:
The global `unbutton` sets the outline-color with the focus selector, so that overrides the default

...and now that I think of it, I should've just done the same and set it on focus - eliminating `!important`. Will do.

jakearchibald commented on src/client/lazy-app/Modal/style.css:
```diff
@@ -0,0 +1,171 @@
+    stroke: black;
+  }
+
+  /* Don't show focus ring on mobile */
+  @media (min-width: 720px) {
+    &:focus {
+      outline-width: 2px;
+      outline-style: solid;
+      outline-offset: 2px;
+    }
+  }
```
Would it be better to use https://developer.mozilla.org/en-US/docs/Web/CSS/:focus-visible ?

jakearchibald commented on src/client/lazy-app/Modal/index.tsx:
```diff
@@ -0,0 +1,136 @@
+                    stroke-width="37"
+                    stroke-linecap="round"
+                  />
+                </svg>
```
Move this to icons?

jakearchibald commented on src/client/lazy-app/Modal/index.tsx:
```diff
@@ -0,0 +1,136 @@
+    return (
+      <dialog
+        ref={linkRef(this, 'dialogElement')}
+        onKeyDown={(e) => this._onKeyDown(e)}
```
You can avoid creating a function on each render by using class fields. See the pattern here https://github.com/GoogleChromeLabs/squoosh/blob/dev/src/shared/prerendered-app/Intro/index.tsx#L282

aryanpingle replied:
Didn't cross my mind that I could change it to an arrow function to preserve the `this` context, will do!

jakearchibald commented on src/client/lazy-app/Modal/index.tsx:
```diff
@@ -0,0 +1,136 @@
+      >
+        {shown && (
+          <Fragment>
+            <header class={style.header}>
```
If you're going to use `<header>` and `<footer>`, then wrap both in an `<article>`. This semantically links the header and footer.

jakearchibald commented on src/client/lazy-app/Modal/index.tsx:
```diff
@@ -0,0 +1,136 @@
+    anim.onfinish = this._closeOnTransitionEnd;
+  }
+
+  private _onKeyDown(e: KeyboardEvent) {
```
Nit: Rename the arg to `event`. Leave minification to the minifier 😄

jakearchibald commented on src/client/lazy-app/Modal/index.tsx:
```diff
@@ -0,0 +1,136 @@
+  };
+
+  showModal() {
+    if (!this.dialogElement || this.dialogElement.open)
+      throw Error('Modal missing / already shown');
```
My preference is that, if the thing is already in the desired state, then just return rather than throwing. Also, try to avoid using both component state & dialog element state. Use one as the source of truth (the component state?)

jakearchibald commented on src/client/lazy-app/Modal/index.tsx:
```diff
@@ -0,0 +1,136 @@
+    if (!this.dialogElement) throw new Error('Modal missing');
+
+    // Set inert by default
+    this.dialogElement.inert = true;
```
Can you talk me through the usage of `inert` in this component?

aryanpingle replied:
I have a feeling it's another instance of obsolete code now that we're using the native `dialog` element, but here's the logic: when hidden, the dialog is made inert to prevent its contents from being `TAB`bed to. Might be a remnant of the time when I didn't use the dialog element and had to ensure the modal wasn't focusable.

jakearchibald replied:
I think so. `<dialog>` content isn't tabbable when it's not-shown.

jakearchibald commented on src/client/lazy-app/Modal/index.tsx:
```diff
@@ -0,0 +1,136 @@
+    // Set inert by default
+    this.dialogElement.inert = true;
+
+    // Prevent events from leaking through the dialog
+    this.dialogElement.onclick = (event) => {
+      event.preventDefault();
+      event.stopImmediatePropagation();
+    };
```
Can you talk me through the reason for this, and why the event is attached this way, rather than the usual Preact way?

aryanpingle replied:
No good reason, I'll refactor it

jakearchibald commented on src/client/lazy-app/Modal/index.tsx:
```diff
@@ -0,0 +1,136 @@
+    // animate modal::backdrop
+    // some browsers don't support ::backdrop, catch those errors
+    try {
+      animateTo(this.dialogElement, [{ opacity: 0 }, { opacity: 1 }], {
+        duration: 250,
+        easing: 'linear',
+        pseudoElement: '::backdrop',
+      });
```
Would it make more sense to use `animateFrom` in cases like this?

aryanpingle replied:
(For browsers that support backdrop) We need fill-mode `forwards` or `both` to keep the backdrop opaque after the animation - but animateFrom uses `back`.
If I edit that part of animateFrom then it essentially becomes animateTo without committing styles, and in that case a try-catch is much more readable IMO.

jakearchibald replied:
I think I get it, but wouldn't the backdrop's default from-CSS opacity be 1, so it wouldn't need a fill-forward?

jakearchibald commented on src/client/lazy-app/Modal/index.tsx:
```diff
@@ -0,0 +1,136 @@
+    // animate modal::backdrop
+    // some browsers don't support ::backdrop, catch those errors
+    try {
+      animateTo(this.dialogElement, [{ opacity: 0 }], {
```
```suggestion
      animateTo(this.dialogElement, { opacity: 0 }, {
```

jakearchibald commented on src/client/lazy-app/Modal/index.tsx:
```diff
@@ -0,0 +1,136 @@
+    try {
+      animateTo(this.dialogElement, [{ opacity: 0 }], {
+        duration: 250,
+        easing: 'linear',
```
```suggestion
        easing: 'ease',
```

jakearchibald commented on src/client/lazy-app/Modal/index.tsx:
```diff
@@ -0,0 +1,136 @@
+    try {
+      animateTo(this.dialogElement, [{ opacity: 0 }, { opacity: 1 }], {
+        duration: 250,
+        easing: 'linear',
```
```suggestion
        easing: 'ease',
```

It looks better 😄

jakearchibald commented on src/client/lazy-app/Modal/ModalHint/style.css:
```diff
@@ -0,0 +1,31 @@
+span.modal-hint {
```
Why require a particular kind of element here? (`span`)

aryanpingle replied:
No reason in particular, just seemed to make it easier to understand

jakearchibald replied:
The good thing about using class names (and CSS modules) is that you can decouple style and semantics. So, if a particular `modal-hint` made sense as an `<aside>`, then that's fine, because the class name can be applied to any element. However, in this case it can only be applied to `<span>`, and I don't think that's a useful restriction.

In general I'd just use class names, and only use tag names in cases where it's justified. Cases like `.icon > svg` are fiiiine, and your use of tag names for dialog content is great, but I think just sticking with a class name is better here.

jakearchibald commented on src/client/lazy-app/Modal/ModalHint/index.tsx:
```diff
@@ -0,0 +1,56 @@
+    return (
+      <span
+        class={style.modalHint}
+        onClick={(event) => {
+          event.preventDefault();
+          event.stopImmediatePropagation();
+        }}
```
Why is this needed?

aryanpingle replied:
Might be obsolete because I prevented click events from _leaking_ out of the dialog, I'll look into it

jakearchibald reviewed (changes requested):
This is coming along nicely!

Where did the icons come from? Do we have the rights to use them?

aryanpingle committed:
Switch from eased to linear animations

aryanpingle committed:
Apply suggestions from code review

Minifying icons using SVGOMG

Co-authored-by: Jake Archibald <jaffathecake@gmail.com>

aryanpingle committed:
Remove obsolete code

aryanpingle committed:
Refactoring

aryanpingle committed:
Add RoundedCrossIcon to icons

aryanpingle committed:
Refactor ModalHint

ModalHint now takes the modal content as its children, and the optional text through props

aryanpingle committed:
Improved icons

aryanpingle committed:
Use `animateFrom` to show modal

aryanpingle committed:
(REMOVE) Update modal demo

jakearchibald commented:
@argyleink @kosamari @surma UX question: Right now the whole label becomes an activator for the modal (see "Auto sample chroma" in the preview). My hunch is that only the icon should activate the modal, and the rest of the label should remain as an activator for the checkbox. Can I get a second opinion?

jakearchibald commented on src/client/lazy-app/icons/index.tsx:
```diff
@@ -11,6 +11,46 @@ const Icon = (props: preact.JSX.HTMLAttributes) => (
+);
+
+export const DiamondStarIcon = (props: preact.JSX.HTMLAttributes) => (
+  <svg viewBox="0 0 48 48" fill="currentColor">
```
```suggestion
  <svg viewBox="0 0 48 48" fill="currentColor" {...props}>
```

Otherwise the parameter is unused.

jakearchibald commented on src/client/lazy-app/icons/index.tsx:
```diff
@@ -11,6 +11,46 @@ const Icon = (props: preact.JSX.HTMLAttributes) => (
+);
+
+export const InfoIcon = (props: preact.JSX.HTMLAttributes) => (
+  <svg
+    viewBox="0 0 520 520"
+    fill="currentColor"
+    xmlns="http://www.w3.org/2000/svg"
+  >
+    <g clip-path="url(#a)">
+      <path
+        fill-rule="evenodd"
+        clip-rule="evenodd"
+        d="M520 0H0v520h520V0ZM230.122 77.186c-8.096 7.7587-12.144 17.3727-12.144 28.842 0 11.469 4.048 21.083 12.144 28.842s17.71 11.638 28.842 11.638c11.132 0 20.578-3.879 28.336-11.638 8.096-7.759 12.144-17.373 12.144-28.842 0-11.4693-4.048-21.0833-12.144-28.842-7.758-8.096-17.204-12.144-28.336-12.144-11.132 0-20.746 4.048-28.842 12.144Zm1.012 106.766h-5.566v199.364c0 10.795 1.856 19.903 5.566 ...
+        fill="currentColor"
+      />
+    </g>
+    <defs>
+      <clipPath id="a">
+        <rect width="520" height="520" rx="70" fill="currentColor" />
+      </clipPath>
+    </defs>
+  </svg>
```
```suggestion
  <svg viewBox="0 0 520 520" fill="currentColor" {...props}>
    <path d="M70 0C31 0 0 31 0 70v380c0 39 31 70 70 70h380c39 0 70-31 70-70V70c0-39-31-70-70-70H70zm189 65c11 0 21 4 28 12 8 8 12 18 12 29s-4 21-12 29c-7 8-17 12-28 12s-21-4-29-12-12-18-12-29 4-21 12-29 18-12 29-12zm-33 119h5c18 0 32 5 44 14 12 10 18 25 18 45v198h-3l-23-3c-7-2-14-5-21-10-6-4-11-10-15-17-4-8-5-17-5-28V184z"/>
  </svg>
```

- Optimised the path
- Added the `{...props}`, otherwise it's an unused argument
- Removed the `xmlns`, it isn't needed in HTML
- Removed the clip-path - it's ok to use clip-paths, but remember that `id="a"` would be going into the document, and IDs have document scope. As it was, the icon would likely weirdly break if there was ever something else with `id="a"` (probably another icon that uses the same pattern).

aryanpingle replied:
Nice catch with the `id`! I guess I've never really considered when the downloaded SVG has attributes that might interfere with the code, I'll keep that in mind

jakearchibald commented on src/client/lazy-app/Modal/style.css:
```diff
@@ -0,0 +1,171 @@
+    outline: var(--medium-light-gray); /* Overwrite default black outline */
+  }
+
+  /* Don't show focus ring on mobile */
```
```suggestion
  /* Only show the focus ring on keyboard interaction */
```

jakearchibald commented on src/client/lazy-app/Modal/index.tsx:
```diff
@@ -0,0 +1,120 @@
+interface Props {
+  icon: VNode;
+  title: string;
+  content: VNode;
```
I think we decided that the content of the model should be the children of the modal?

jakearchibald commented on src/client/lazy-app/Modal/index.tsx:
```diff
@@ -0,0 +1,120 @@
+        onClick={(event) => {
+          // Prevent clicks from leaking out of the dialog
+          event.preventDefault();
+          event.stopImmediatePropagation();
```
I don't think it needs to stop immediate propogation?

jakearchibald commented on src/client/lazy-app/Modal/index.tsx:
```diff
@@ -0,0 +1,120 @@
+        <article class={style.article}>
+          <header class={style.header}>
+            <span class={style.modalIcon}>{icon}</span>
+            <span class={style.modalTitle}>{title}</span>
```
This should be a heading element. `<h1>` seems fine.

jakearchibald commented on src/client/lazy-app/Modal/index.tsx:
```diff
@@ -0,0 +1,120 @@
+            <span class={style.modalIcon}>{icon}</span>
+            <span class={style.modalTitle}>{title}</span>
+            <button class={style.closeButton} onClick={() => this._hideModal()}>
+              <RoundedCrossIcon></RoundedCrossIcon>
```
```suggestion
              <RoundedCrossIcon/>
```

jakearchibald commented on src/client/lazy-app/Modal/index.tsx:
```diff
@@ -0,0 +1,120 @@
+  render({ title, icon, content }: Props, {}: State) {
+    return (
+      <dialog
+        class={style['modalDialog']}
```
```suggestion
        class={style.modalDialog}
```

jakearchibald commented on src/client/lazy-app/Modal/index.tsx:
```diff
@@ -0,0 +1,120 @@
+    if (event.key === 'Escape') {
+      this._hideModal();
+      event.preventDefault();
+      event.stopImmediatePropagation();
```
I don't think `stopImmediatePropagation` is needed

jakearchibald commented on src/client/lazy-app/Modal/index.tsx:
```diff
@@ -0,0 +1,120 @@
+        pseudoElement: '::backdrop',
+      });
+    } catch (e) {}
+    anim.onfinish = this._closeOnTransitionEnd;
```
```suggestion
    anim.finished.then(() => this._closeOnTransitionEnd());
```

Using the promise makes it clearer that we only expect this to happen once. Also, it avoids passing an event into a function that doesn't expect it. https://jakearchibald.com/2021/function-callback-risks/

jakearchibald commented on src/client/lazy-app/Modal/index.tsx:
```diff
@@ -0,0 +1,120 @@
+  }
+
+  private _hideModal() {
+    if (!this.dialogElement) throw Error('Modal missing / hidden');
```
I think it's ok to assume `dialogElement` is there, since this component never presents a state where it isn't. Other instances of this can be skipped too.

jakearchibald commented on src/client/lazy-app/Modal/ModalHint/index.tsx:
```diff
@@ -0,0 +1,57 @@
+
+interface Props {
+  modalTitle: string;
+  text?: string;
```
Let's change this to "label" or something that better describes its usage.

jakearchibald commented on src/client/lazy-app/Modal/ModalHint/index.tsx:
```diff
@@ -0,0 +1,57 @@
+          onClick={this.onclick}
+          title="Learn more"
+        >
+          <InfoIcon></InfoIcon>
```
```suggestion
          <InfoIcon/>
```

jakearchibald commented on src/client/lazy-app/Modal/ModalHint/index.tsx:
```diff
@@ -0,0 +1,57 @@
+          // When the button is clicked, the event starts bubbling up
+          // which might cause unexpected behaviour
+          event.preventDefault();
+          event.stopImmediatePropagation();
```
I don't think `stopImmediatePropagation` is needed.

jakearchibald commented on src/client/lazy-app/Modal/ModalHint/index.tsx:
```diff
@@ -0,0 +1,57 @@
+
+  private onclick = (event: Event) => {
+    if (!this.modalComponent)
+      throw new Error('ModalHint is missing a modalComponent');
```
This can probably be skipped, as it's guaranteed to be there.

jakearchibald reviewed (changes requested):
Coming along nicely!

argyleink commented:
> UX question: Right now the whole label becomes an activator for the modal (see "Auto sample chroma" in the preview). My hunch is that only the icon should activate the modal, and the rest of the label should remain as an activator for the checkbox. Can I get a second opinion?

Confirm, label clicks are typically activators for an associated checkbox, plus this is also how Squoosh works today, using labels as activators.

I agree with Jake, the icon should isolate the interaction to learn more from a modal, both because users expect labels to activate checkboxes on the web, but also because that's the UX that users currently expect from the label interactions.

Same goes for the nested link in the mozJPEG quality modal, a link should be a link and a button to invoke a modal should be a button to invoke a modal.

Button was a good element choice to wrap the icon, as this keeps it in keyboard tab flow 👍🏻

aryanpingle committed:
Compress SVGs

* Optimized SVG paths
* Removed unnecessary attributes

Co-authored-by: Jake Archibald <jaffathecake@gmail.com>
